mjpc/planners/sample_gradient/planner.cc

```diff
@@ -0,0 +1,652 @@
+        s.candidate_policy[i].representation =
+            s.resampled_policy.representation;
+
+        // noisy nominal policy
+        if (i > idx_nominal) s.AddNoiseToPolicy(i);
+      }
+
+      // ----- rollout sample policy ----- //
+
+      // policy
+      auto sample_policy_i = [&candidate_policy = s.candidate_policy, &i](
+                                 double* action, const double* state,
+                                 double time) {
+        candidate_policy[i].Action(action, state, time);
+      };
+
+      // policy rollout
+      s.trajectory[i].Rollout(
+          sample_policy_i, task, model, s.data_[ThreadPool::WorkerId()].get(),
+          state.data(), time, mocap.data(), userdata.data(), horizon);
+    });
+  }
+  pool.WaitCount(count_before + num_trajectory);
+  pool.ResetCount();
+}
+
+// compute candidate trajectories
+void SampleGradientPlanner::GradientCandidates(int num_trajectory,
+                                               int num_gradient, int horizon,
+                                               ThreadPool& pool) {
+  if (num_gradient < 1) return;
+
+  // number of parameters
+  int num_parameters = resampled_policy.num_parameters;
+  int num_spline_points = resampled_policy.num_spline_points;
+
+  // cache old gradient
+  mju_copy(gradient_previous.data(), gradient.data(), num_parameters);
+
+  // -- compute approximate gradient -- //
+  // average return
+  int num_noisy = num_trajectory - num_gradient;
+  double avg_return = 0.0;
+  for (int i = 0; i < num_noisy; i++) {
+    avg_return += trajectory[i].total_return;
+  }
+  avg_return /= num_noisy;
+
+  // gradient
+  std::fill(gradient.begin(), gradient.end(), 0.0);
+  for (int i = 1; i < num_noisy; i++) {
+    double* noisei = noise.data() + i * (model->nu * kMaxTrajectoryHorizon);
+    mju_addToScl(gradient.data(), noisei,
+                 (trajectory[i].total_return - avg_return) / (num_noisy - 1),
+                 num_parameters);
+  }
+
+  // normalize gradient
+  // TODO(taylor): should we normalize?
+  mju_normalize(gradient.data(), num_parameters);
+
+  // compute step sizes along gradient
+  std::vector<double> step_size(num_gradient);
+  LogScale(step_size.data(), gradient_max_step_size, gradient_min_step_size,
+           num_gradient);
+
+  // gradient filter gf * grad + (1 - gf) * grad_prev
+  double gradient_filter = gradient_filter_;
+
+  // compute candidate policies along gradient direction
+  // these candidates will be evaluated at the next planning iteration
+  for (int i = num_noisy; i < num_trajectory; i++) {
+    // copy nominal policy
+    candidate_policy[i].CopyFrom(resampled_policy, num_spline_points);
+    candidate_policy[i].representation = resampled_policy.representation;
+
+    // scaling
+    // TODO(taylor): scale by num_parameters?
+    // TODO(taylor): divide by reward std dev?
+    double scaling = step_size[i - num_noisy] / noise_exploration;
```

Comment:
okay, you're doing it here, that's fine, but I think it's `* noise_exploration * num_parameters`. Certainly division doesn't make sense to me.

Reply:
I based it off [algorithm 1, line 5](https://arxiv.org/pdf/1703.03864.pdf).

Reply:
<img width="667" alt="image" src="https://github.com/google-deepmind/mujoco_mpc/assets/12521315/1888fc9d-6dc8-40fd-8ce6-d337c2abfa07">